### ISSUE TITLE
Syntax error when using DBT Jinja variable with Snowflake + DBT

### ISSUE THREAD

**Issue by Ben-A-Canva** [sqlfluff/sqlfluff#2235]
When linting a file containing the following section of code, the sqlfluff slicer includes the `alternate_code` segment which breaks the dbt compile run.

```
{%- set get_source_maximum_query %}
  select
    max(_partition_key) as max_partition_key
  from snowflake_db.{{ target.name }}__raw._test_table
{%- endset %}
```
The generated SQL is:
```
  select
    max(_partition_key) as max_partition_key
  from snowflake_db.659cfb6f9b26429aa2cf83619d3bf5f9 dev\x00__raw._test_table
```
Resulting in ```SQL compilation error: syntax error line 3 at position 21 unexpected '659'.``` In Snowflake's query history.

From debugging, i've found that the templater slicer is adding in the `alternate_code` which results in the above query.

## Expected Behaviour
The generated sql should be 
```
  select
    max(_partition_key) as max_partition_key
  from snowflake_db.dev__raw._test_table
```

## Dialect
Snowflake

## Version
sqlfluff, version 0.9.0
dbt version 1.0.0

## Configuration
```
[sqlfluff]
verbose = 2
nocolor = False
dialect = snowflake
templater = dbt
exclude_rules = L031,L034,L022,L051 
#L025,
recurse = 10
output_line_length = 500
runaway_limit = 10
ignore_templated_areas = True
encoding = 'utf-8'
format = 'human'
# Comma separated list of file extensions to lint.

# NB: This config will only apply in the root folder.
sql_file_exts = .sql,.sql.j2,.dml,.ddl

[sqlfluff:templater:dbt]
project_dir = src/app/transform
profiles_dir = src/app
profile = app

[sqlfluff:indentation]
indented_joins = False
template_blocks_indent = True

[sqlfluff:templater]
unwrap_wrapped_queries = False

[sqlfluff:templater:jinja:macros]
#Macros provided as builtins for dbt projects
dbt_ref = {% macro ref(model_ref) %}{{model_ref}}{% endmacro %}
dbt_source = {% macro source(source_name, table) %}{{source_name}}_{{table}}{% endmacro %}
dbt_config = {% macro config() %}{% for k in kwargs %}{% endfor %}{% endmacro %}
dbt_var = {% macro var(variable) %}item{% endmacro %}
dbt_is_incremental = {% macro is_incremental() %}False{% endmacro %}

# Some rules can be configured directly from the config common to other rules.
[sqlfluff:rules]
tab_space_size = 2
max_line_length = 100
indent_unit = space
comma_style = trailing
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all


# Some rules have their own specific config.
[sqlfluff:rules:L003]
indent_unit = space
tab_space_size = 2

[sqlfluff:rules:L010]  # Keywords
capitalisation_policy = lower

[sqlfluff:rules:L014]
extended_capitalisation_policy = lower
unquoted_identifiers_policy = all

[sqlfluff:rules:L016]
ignore_comment_lines = True

[sqlfluff:rules:L029]  # Keyword identifiers
unquoted_identifiers_policy = all

[sqlfluff:rules:L030]  # Function names
capitalisation_policy = lower

[sqlfluff:rules:L038]
select_clause_trailing_comma = forbid

[sqlfluff:rules:L040]  # Null & Boolean Literals
capitalisation_policy = lower

[sqlfluff:rules:L042]
# By default, allow subqueries in from clauses, but not join clauses.
forbid_subquery_in = both

[sqlfluff:rules:L047]  # Consistent syntax to count all rows
prefer_count_0 = False
prefer_count_1 = False
```

**Comment by jpy-git** [comment]
@barrywhart may be of interest on your current corruption fix tear 😄

**Comment by barrywhart** [comment]
This seems to be the part that causes the failure:
```
{{ target.name }}
```

I'm thinking this area of the code isn't handling the case where a `{% set %}` block uses template substitution.

**Comment by tunetheweb** [comment]
I was thinking it was more to do with #2236 and referencing an inset variable in the mean time?

**Comment by barrywhart** [comment]
This particular Jinja macro is using `target`, but the failure here would happen regardless of what's inside the `{{ }}`. So the user may also be facing #2236, but there's definitely a bug in the template-related components.

I was just looking into the templating aspect. This may be a tricky bug, or it may be a really easy one. I'll create a PR with an experimental change for the user to try. That'll be up in a few minutes.

**Comment by barrywhart** [comment]
@Ben-A-Canva: I took a quick look at this issue, and I  _may_ have a fix for it. Could you test out this PR? #2240

It makes some changes to the underlying templating components of SQLFluff.

Let me know if it fixes your issue. If it _doesn't_ fix your issue, can you provide a more complete test case? For now, I'm testing with the following model file:
```
{% set get_source_maximum_query %}
  select
    max(_partition_key) as max_partition_key
  from snowflake_db.{{ target.name }}__raw._test_table
{% endset %}

{{ get_source_maximum_query }}
```

And it renders as:
```



  select
    max(_partition_key) as max_partition_key
  from snowflake_db.dev__raw._test_table


```

which doesn't have the weird, long hex number you were seeing in your rendered SQL. 🤞

**Comment by barrywhart** [comment]
After further testing, my quick fix attempt is flawed. I'm digging further into it. A more complete test case would still be helpful. 💯

**Comment by Ben-A-Canva** [comment]
Hi Barry, thanks for taking a look at this so quickly :) I tested out that PR but I'm still getting the hex numbers :( 
After a bit more debugging i think the issue might stem from the `run_query` macro. Here's a test case that fails for me:
```{{
  config(
    alias='test_model',
    copy_grants=true,
    enabled=true,
    materialized='table',
    tags=[
      'snapshot',
    ],
  )
}}
{%- set get_source_maximum_query %}
  select
    max(_partition_key) as max_partition_key
  from snowflake_db.{{ target.target_name }}__raw._test_table
{%- endset %}
{%- if execute %}
{%- set get_source_maximum_query_result = run_query(get_source_maximum_query) %}
{%- endif %}
select
  1 as value
```

If i comment out the line with `run_query` everything seems to work. It also seems like `select max...` is run twice in snowflake. The first time correctly then the second time with `long hex number`:
![image](https://user-images.githubusercontent.com/81595098/148301894-73e92da9-c824-49d4-a394-85d83ba424b9.png)

**Comment by barrywhart** [comment]
I think there are two issues here:
* Incorrect handling when a `{% set %}` block contains a `{{ }}` block, which is this line of your test case: `  from snowflake_db.{{ target.target_name }}__raw._test_table`.
* In order to "understand" the structure of a Jinja or dbt template, SQLFluff needs to build and execute an alternate version of the template (with the hex values you've noticed).

The first item above should be fixed now in the updated PR.

The second issue causes problems with `run_query()`, because the alternate version of the template replaces the usual values in `get_source_maximum_query` with hex values. At this time, I don't know of a way to fix this in SQLFluff. I'm not that knowledgeable about dbt, but are there potentially other ways to achieve the same thing that avoid invoking `{% set %}` with the output of a previous use of `{% set %}`?

In other words, the challenging bit involves this pattern:
```
{%- set my_variable_1 %}
SOMETHING SOMETHING
{%- endset %}
...
{%- set my_variable2 = do_something_with(my_variable_1) %}
```

**Comment by barrywhart** [comment]
I wonder if there's a way to configure things so during SQLFluff linting, you use different config values than in production. I'm thinking your model file, when run in "lint mode", would execute differently, skipping the steps that don't work in SQLFluff, e.g. `run_query()`.

**Comment by jpy-git** [comment]
@barrywhart does #2240 close this?

**Comment by barrywhart** [comment]
@jpy-git: This is the issue I mentioned yesterday where there are two things happening. See this [comment](https://github.com/sqlfluff/sqlfluff/issues/2235#issuecomment-1006174928) above for more details. My PR fixes the first thing. The second thing is not fixed, and really can't be fixed with the current design of `JinjaTracer`. My advice would either be to use flags to skip the problematic code during linting, or to move the code to separate macro file(s).

I'll go ahead and close the issue, since we don't plan to make any more changes to SQLFluff related to this. At some point, we may want to consider documenting these limitations and how to work around them.
